transaction-view/src/resolved_transaction_view.rs

```diff
@@ -0,0 +1,482 @@
+    ) -> Vec<bool> {
+        // Build account keys so that we can iterate over and check if
+        // an address is writable.
+        let account_keys = AccountKeys::new(view.static_account_keys(), Some(resolved_addresses));
+
+        let mut is_writable_cache = Vec::with_capacity(account_keys.len());
+        let num_static_accounts = usize::from(view.num_static_account_keys());
+        let num_writable_lookup_accounts = usize::from(view.total_writable_lookup_accounts());
+        let num_signed_accounts = usize::from(view.num_required_signatures());
+        let num_unsigned_accounts = num_static_accounts.wrapping_sub(num_signed_accounts);
+        let num_writable_unsigned_accounts =
+            num_unsigned_accounts.wrapping_sub(usize::from(view.num_readonly_unsigned_accounts()));
+        let num_writable_signed_accounts =
+            num_signed_accounts.wrapping_sub(usize::from(view.num_readonly_signed_accounts()));
+
+        let is_upgradable_loader_present = account_keys
+            .iter()
+            .any(|key| bpf_loader_upgradeable::ID == *key);
+
+        for (index, key) in account_keys.iter().enumerate() {
+            let is_requested_write = {
+                // If the account is a resolved address, check if it is writable.
+                if index >= num_static_accounts {
+                    let loaded_address_index = index.wrapping_sub(num_static_accounts);
+                    loaded_address_index < num_writable_lookup_accounts
+                } else if index >= num_signed_accounts {
+                    let unsigned_account_index = index.wrapping_sub(num_signed_accounts);
+                    unsigned_account_index < num_writable_unsigned_accounts
+                } else {
+                    index < num_writable_signed_accounts
+                }
+            };
+
+            // If the key is reserved it cannot be writable.
+            is_writable_cache.push(is_requested_write && !reserved_account_keys.contains(key));
+        }
+
+        // If the upgradable loader is not present and a key is called as a program
+        // it is not writable.
+        // If the upgradable loader is present, then there is no point to loop.
+        // Looping over the instructions is more efficient than looping over each account
+        // and checking `is_invoked` which loops over each instruction internally.
+        if !is_upgradable_loader_present {
+            for ix in view.instructions_iter() {
+                is_writable_cache[usize::from(ix.program_id_index)] = false;
+            }
+        }
+
+        is_writable_cache
+    }
+
+    fn num_readonly_accounts(&self) -> usize {
+        usize::from(
+            self.view
+                .total_readonly_lookup_accounts()
+                .wrapping_add(u16::from(self.view.num_readonly_signed_accounts()))
+                .wrapping_add(u16::from(self.view.num_readonly_unsigned_accounts())),
+        )
```

Comment:
Technically it's impossible of course for this to overflow a u16 but it feels slightly more correct to using wrapping addition over usize rather u16 here.

Reply:
56be3995e0512dd8a8418fdbcfa228fedfcd1263